app/grandchallenge/components/models.py

```diff
@@ -2087,57 +2094,95 @@ def assign_permissions(self):
 
 
 class CIVForObjectMixin:
-    def create_civ(self, *, ci_slug, new_value, user=None):
+    def create_civ(self, *, ci_slug, new_value, user=None, linked_task=None):
         ci = ComponentInterface.objects.get(slug=ci_slug)
         try:
-            current_civ = self.values.filter(interface=ci).get()
+            current_civ = (
+                getattr(self, self.base_object.civ_set_lookup)
+                .filter(interface=ci)
+                .get()
+            )
         except ObjectDoesNotExist:
             current_civ = None
         except MultipleObjectsReturned as e:
             raise e
 
         if ci.is_json_kind and not ci.requires_file:
             return self.create_civ_for_value(
-                ci=ci, current_civ=current_civ, new_value=new_value
+                ci=ci,
+                current_civ=current_civ,
+                new_value=new_value,
+                linked_task=linked_task,
             )
         elif ci.is_image_kind:
             return self.create_civ_for_image(
-                ci=ci, current_civ=current_civ, new_value=new_value, user=user
+                ci=ci,
+                current_civ=current_civ,
+                new_value=new_value,
+                user=user,
+                linked_task=linked_task,
             )
         elif ci.requires_file:
             return self.create_civ_for_file(
-                ci=ci, current_civ=current_civ, new_value=new_value
+                ci=ci,
+                current_civ=current_civ,
+                new_value=new_value,
+                linked_task=linked_task,
             )
         else:
             NotImplementedError(f"CIV creation for {ci} not handled.")
 
-    def create_civ_for_value(self, *, ci, current_civ, new_value):
+    def create_civ_for_value(self, *, ci, current_civ, new_value, linked_task):
         current_value = current_civ.value if current_civ else None
-        civ = ComponentInterfaceValue(interface=ci, value=new_value)
+        if new_value is not None:
+            civ, created = ComponentInterfaceValue.objects.get_or_create(
+                interface=ci, value=new_value
+            )
+        else:
+            civ = ComponentInterfaceValue(interface=ci)
         if current_value != new_value or (
             current_civ is None and new_value is None
         ):
             try:
                 civ.full_clean()
                 civ.save()
-                self.values.remove(current_civ)
-                self.values.add(civ)
+                getattr(self, self.base_object.civ_set_lookup).remove(
+                    current_civ
+                )
+                getattr(self, self.base_object.civ_set_lookup).add(civ)
             except ValidationError as e:
                 if new_value:
-                    raise e
+                    if hasattr(self, "error_message"):
+                        self.status = self.CANCELLED
+                        self.error_message = format_validation_error_message(e)
```

Comment:
Are these attributes guaranteed to exist and be db fields? This is only a mixin, does it just happen to work for all users of the mixin? Why are you not using the update status method?

Reply:
Setting the status will only work for Jobs, not archive items and display sets. This is why I checked for the existing of the error_message attribute, but on second thought I think being explicit and checking if `isinstance(self, Job)` is better and I will also use `update_status`.

Reply:
Maybe using `hasattr(self, "update_status")` rather than checking for a particular class? Or add it to the base class and leave it as `pass` by default?